README.md
# Name Service Switch (NSS) module for Juju (Proof of Concept).

This Name Service Switch (NSS) module provides address resolution for
Juju encoded hostnames.

## Supporting Juju charms

Some charms require a hostname in lieu of an IP address to work
properly. To support this you can use ```network-get
--primary-hostname``` in a charm hook to get an always-resolvable
hostname. Address resolution is provided by this plugin.

The encoding of the name is *TBD* but this proof of concept currently
supports the following patterns:

* IPv4

	juju-ip-172-31-38-207

* IPv6

	juju-ip-2001-470-1f1d-8d8-c0db-9559-9417-2416

IPv6 names must be fully expanded (i.e., no collapsed 0's (zeros)).

### Juju Charm Example

	$ network-get --primary-hostname
	juju-ip-192-168-20-1

### General Address Resolution

	$ getent hosts juju-ip-192-168-20-1
	192.168.20.1    juju-ip-192-168-20-1

	$ ping -c 4 juju-ip-192-168-20-1
	PING juju-ip-192-168-20-1 (192.168.20.1) 56(84) bytes of data.
	64 bytes from 192.168.20.1: icmp_seq=1 ttl=64 time=0.069 ms
	64 bytes from 192.168.20.1: icmp_seq=2 ttl=64 time=0.052 ms
	64 bytes from 192.168.20.1: icmp_seq=3 ttl=64 time=0.048 ms
	64 bytes from 192.168.20.1: icmp_seq=4 ttl=64 time=0.048 ms

## Limitations

### Reverse lookup needs to be handled by the host

The module itself cannot be authoritative for arbitrary IP addresses,
particularly if the module is listed before the ```dns``` entry. It
can, however, rely on the host to do reverse lookup.

#### Reverse lookup on AWS

	$ hostname
	ip-172-31-0-139

	$ network-get --primary-hostname
	juju-ip-172-31-0-139

	$ getent hosts juju-ip-172-31-0-139
	172.31.0.139    juju-ip-172-31-0-139

	# Reverse lookup
	$ dig -x 172.31.0.139 +short
	ip-172-31-0-139.ec2.internal.

	# Forward lookup
	$ dig ip-172-31-0-139.ec2.internal. +short
	172.31.0.139

#### Reverse lookup on CE

Works and behaves the same way as AWS does.

#### Reverse lookup on Azure

Reverse lookup does **not** yield a result.

### Only IPv4 or IPv6 for a single hostname

Because the IP address is encoded in the name it is not (currently)
possible to get both an IPv4 and IPv6 address associated for one
hostname. This can be fixed by encoding both addresses in the name
(but it doesn't read terribly well).

   $ juju-ip-172-31-38-207-2001-470-1f1d-8d8-c0db-9559-9417-2416

# Building from source

Please read the [HACKING](HACKING.md) companion to this file.


NEWS.md
## Version 0.1

04 Sep 2016

* Proof of Concept


tests/gethostbyname.c
/* Copyright 2016 Canonical Ltd.
 *
 * Licensed under the AGPLv3, see LICENSE file for details.
 *
 * Name Service Switch module for resolving Juju host names.
 */

#include <check.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <arpa/inet.h>
#include "nss-juju.h"

#define NELEMENTS(A) ((sizeof((A)) / sizeof((A))[0]))

#define GETHOSTBYNAME_VARIABLES			\
	struct hostent	 hostent;		\
	char		 buffer[1024];		\
	int		 err;			\
	int		 h_err			\

#define GETHOSTBYNAME2_VARIABLES		\
	struct hostent	 hostent;		\
	char		 buffer[1024];		\
	int		 err;			\
	int		 h_err

#define GETHOSTBYNAME3_VARIABLES		\
	struct hostent	 hostent;		\
	char		 buffer[1024];		\
	int		 err;			\
	int		 h_err;			\
	int32_t		 ttl;			\
	char		*canon = NULL

static const char *invalid_names[] = {
	"juju",
	"juju-ip",
	"juju-ip-",
	"juju-ip-192",
	"juju-ip-192-168",
	"juju-ip-192-168-1",
	"juju-ip-192-168-1-1 ",
	"juju-ip-192-168-1-1-2 ",
	"juju-ip-172-16-30-1 trailing chars",
	"juju-ip-172-16-30-1-trailing chars",
	"juju-ip-172-16-30-11-trailing chars",
	"juju-ip-172-16-30-123-trailing chars",
	"juju-ip-255-255-255-256",
	"juju-ip-172.16.30.1234",
	"juju-ip-256-256-256-256",
	"leading chars juju-ip-255-255-255-255",
	"juju-ip-256-256-256-256",
	"juju-ip-192-168-200-1 ",
	"leading chars juju-ip-192-168-200-1",
	"juju-ip-10-11-12-256",
	"juju-ip-2001-0470-1f1d-08d8-c0db-9559-9417",
};

struct test_case {
	const char *name;
	const char *expected_ip_str;
	int af;
};

static struct test_case valid_names[] = {
	{ "juju-ip-1-1-1-1",
	  "1.1.1.1",
	  AF_INET },
	{ "juju-ip-10-10-10-10",
	  "10.10.10.10",
	  AF_INET },
	{ "juju-ip-192-168-199-101",
	  "192.168.199.101",
	  AF_INET },
	{ "juju-ip-2001-470-1f1d-8d8-c0db-9559-9417-2416",
	  "2001:470:1f1d:8d8:c0db:9559:9417:2416",
	  AF_INET6 },
	{ "juju-ip-2001-1470-1f1d-f8d8-c0db-9559-9417-2416",
	  "2001:1470:1f1d:f8d8:c0db:9559:9417:2416",
	  AF_INET6 }
};

static size_t nhosts(struct hostent *h)
{
	size_t nhosts = 0;
	char **pptr;
	for (pptr = h->h_addr_list; *pptr != NULL; pptr++, nhosts++) {}
	return nhosts;
}

static const char *hostent_str(struct hostent *h, size_t n, char *buf)
{
	return inet_ntop(h->h_addrtype, h->h_addr_list[n], buf, INET6_ADDRSTRLEN);
}

static enum nss_status _nss_juju_gethostbyname_wrapper(
	int		  version,
	const char	 *name,
	int		  af,
	struct hostent	 *hostent,
	char		 *buffer,
	size_t		  buflen,
	int		 *errnop,
	int		 *herrnop,
	int32_t		 *ttl,
	char		**canonp)
{
	*canonp = NULL;
	*ttl = -1;

	switch (version) {
	case 3:
		return _nss_juju_gethostbyname3_r(
			name,
			af,
			hostent,
			buffer,
			buflen,
			errnop,
			herrnop,
			ttl,
			canonp);
	case 2:
		return _nss_juju_gethostbyname2_r(
			name,
			af,
			hostent,
			buffer,
			buflen,
			errnop,
			herrnop);
	default:
		return _nss_juju_gethostbyname_r(
			name,
			hostent,
			buffer,
			buflen,
			errnop,
			herrnop);
	}
}

START_TEST(test_null_name)
{
	const char	*name	= NULL;
	struct hostent	 hostent;
	char		*buffer = NULL;
	int		 err;
	int		 h_err;
	int32_t		 ttl;
	char		*canonp = NULL;
	enum nss_status	 res;

	res = _nss_juju_gethostbyname_wrapper(
		_i,
		name,
		AF_UNSPEC,
		&hostent,
		buffer,
		0,
		&err,
		&h_err,
		&ttl,
		&canonp);

	fail_unless(res == NSS_STATUS_UNAVAIL);
	fail_unless(err == EINVAL);
}
END_TEST

START_TEST(test_zerolen_name)
{
	const char	*name	= "";
	struct hostent	 hostent;
	char		*buffer = NULL;
	int		 err;
	int		 h_err;
	int32_t		 ttl;
	char		*canonp = NULL;
	enum nss_status	 res;

	res = _nss_juju_gethostbyname_wrapper(
		_i,
		name,
		AF_UNSPEC,
		&hostent,
		buffer,
		0,
		&err,
		&h_err,
		&ttl,
		&canonp);

	fail_unless(res == NSS_STATUS_NOTFOUND);
	fail_unless(err == ENOENT);
}
END_TEST

START_TEST(test_invalid_name_gethostbyname)
{
	GETHOSTBYNAME_VARIABLES;

	enum nss_status res = _nss_juju_gethostbyname_r(
		invalid_names[_i],
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err);

	fail_unless(res == NSS_STATUS_NOTFOUND);
	fail_unless(err == EADDRNOTAVAIL);
}
END_TEST

START_TEST(test_invalid_name_gethostbyname2)
{
	GETHOSTBYNAME2_VARIABLES;

	enum nss_status res = _nss_juju_gethostbyname2_r(
		invalid_names[_i],
		AF_UNSPEC,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err);

	fail_unless(res == NSS_STATUS_NOTFOUND);
	fail_unless(err == EADDRNOTAVAIL);
}
END_TEST

START_TEST(test_invalid_name_gethostbyname3)
{
	GETHOSTBYNAME3_VARIABLES;

	enum nss_status res = _nss_juju_gethostbyname3_r(
		invalid_names[_i],
		AF_UNSPEC,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err,
		&ttl,
		&canon);

	fail_unless(res == NSS_STATUS_NOTFOUND);
	fail_unless(err == EADDRNOTAVAIL);
}
END_TEST

START_TEST(test_valid_ipv4_name_but_null_buffer)
{
	const char	*name	= "juju-ip-192-168-20-1";
	struct hostent	 hostent;
	char		*buffer = NULL;
	int		 err;
	int		 h_err;
	int32_t		 ttl;
	char		*canonp = NULL;
	enum nss_status	 res;

	res = _nss_juju_gethostbyname_wrapper(
		_i,
		name,
		AF_UNSPEC,
		&hostent,
		buffer,
		0,
		&err,
		&h_err,
		&ttl,
		&canonp);

	fail_unless(res == NSS_STATUS_UNAVAIL);
	fail_unless(err == EINVAL);
}
END_TEST

START_TEST(test_valid_ipv6_name_but_null_buffer)
{
	const char	*name	= "juju-ip-2001-470-1f1d-8d8-c0db-9559-9417-2416";
	struct hostent	 hostent;
	char		*buffer = NULL;
	int		 err;
	int		 h_err;
	int32_t		 ttl;
	char		*canonp = NULL;
	enum nss_status	 res;

	res = _nss_juju_gethostbyname_wrapper(
		_i,
		name,
		AF_UNSPEC,
		&hostent,
		buffer,
		0,
		&err,
		&h_err,
		&ttl,
		&canonp);

	fail_unless(res == NSS_STATUS_UNAVAIL);
	fail_unless(err == EINVAL);
}
END_TEST

START_TEST(test_valid_name_gethostbyname)
{
	GETHOSTBYNAME_VARIABLES;
	char ipbuf[INET6_ADDRSTRLEN];

	enum nss_status res = _nss_juju_gethostbyname_r(
		valid_names[_i].name,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err);

	fail_unless(res == NSS_STATUS_SUCCESS);
	fail_unless(err == 0);
	fail_unless(h_err == 0);
	fail_unless(*hostent.h_aliases == NULL);
	fail_unless(nhosts(&hostent) == 1);
	fail_unless(hostent.h_addrtype == valid_names[_i].af);
	ck_assert_str_eq(hostent.h_name, valid_names[_i].name);
	ck_assert_str_eq(hostent_str(&hostent, 0, ipbuf), valid_names[_i].expected_ip_str);
}
END_TEST

START_TEST(test_valid_name_gethostbyname2)
{
	GETHOSTBYNAME2_VARIABLES;
	char ipbuf[INET6_ADDRSTRLEN];

	enum nss_status res = _nss_juju_gethostbyname2_r(
		valid_names[_i].name,
		valid_names[_i].af,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err);

	fail_unless(res == NSS_STATUS_SUCCESS);
	fail_unless(err == 0);
	fail_unless(h_err == 0);
	fail_unless(*hostent.h_aliases == NULL);
	fail_unless(nhosts(&hostent) == 1);
	fail_unless(hostent.h_addrtype == valid_names[_i].af);
	ck_assert_str_eq(hostent.h_name, valid_names[_i].name);
	ck_assert_str_eq(hostent_str(&hostent, 0, ipbuf), valid_names[_i].expected_ip_str);
}
END_TEST

START_TEST(test_valid_name_gethostbyname3)
{
	GETHOSTBYNAME3_VARIABLES;
	char ipbuf[INET6_ADDRSTRLEN];

	enum nss_status res = _nss_juju_gethostbyname3_r(
		valid_names[_i].name,
		valid_names[_i].af,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err,
		&ttl,
		&canon);

	fail_unless(res == NSS_STATUS_SUCCESS);
	fail_unless(err == 0);
	fail_unless(h_err == 0);
	fail_unless(*hostent.h_aliases == NULL);
	fail_unless(nhosts(&hostent) == 1);
	fail_unless(hostent.h_addrtype == valid_names[_i].af);
	ck_assert_str_eq(hostent.h_name, valid_names[_i].name);
	ck_assert_str_eq(hostent_str(&hostent, 0, ipbuf), valid_names[_i].expected_ip_str);
	fail_unless(ttl == 0);
	ck_assert_str_eq(canon, valid_names[_i].name);
}
END_TEST

START_TEST(test_valid_name_with_small_buffer)
{
	GETHOSTBYNAME3_VARIABLES;
	char ipbuf[INET6_ADDRSTRLEN];

	enum nss_status res = _nss_juju_gethostbyname_wrapper(
		_i,
		valid_names[_i].name,
		valid_names[_i].af,
		&hostent,
		buffer,
		1,
		&err,
		&h_err,
		&ttl,
		&canon);

	fail_unless(res == NSS_STATUS_TRYAGAIN);
	fail_unless(err == ERANGE);

	res = _nss_juju_gethostbyname_wrapper(
		_i,
		valid_names[_i].name,
		valid_names[_i].af,
		&hostent,
		buffer,
		sizeof(buffer),
		&err,
		&h_err,
		&ttl,
		&canon);

	fail_unless(res == NSS_STATUS_SUCCESS);
	fail_unless(err == 0);
	fail_unless(h_err == 0);
	fail_unless(*hostent.h_aliases == NULL);
	fail_unless(nhosts(&hostent) == 1);
	fail_unless(hostent.h_addrtype == valid_names[_i].af);
	ck_assert_str_eq(hostent.h_name, valid_names[_i].name);
	ck_assert_str_eq(hostent_str(&hostent, 0, ipbuf), valid_names[_i].expected_ip_str);
	switch (_i) {
	case 3:
		fail_unless(ttl == 0);
		ck_assert_str_eq(canon, valid_names[_i].name);
		break;
	}
}
END_TEST

static Suite *libnss_juju_suite(void)
{
	Suite *s = suite_create("nss-juju");
	TCase *tc = tcase_create("gethostbyname");
	tcase_add_loop_test(tc, test_null_name, 1, 4);
	tcase_add_loop_test(tc, test_zerolen_name, 1, 4);
	tcase_add_loop_test(tc, test_invalid_name_gethostbyname, 0, NELEMENTS(invalid_names));
	tcase_add_loop_test(tc, test_invalid_name_gethostbyname2, 0, NELEMENTS(invalid_names));
	tcase_add_loop_test(tc, test_invalid_name_gethostbyname3, 0, NELEMENTS(invalid_names));
	tcase_add_loop_test(tc, test_valid_ipv4_name_but_null_buffer, 1, 4);
	tcase_add_loop_test(tc, test_valid_ipv6_name_but_null_buffer, 1, 4);
	tcase_add_loop_test(tc, test_valid_name_gethostbyname, 0, NELEMENTS(valid_names));
	tcase_add_loop_test(tc, test_valid_name_gethostbyname2, 0, NELEMENTS(valid_names));
	tcase_add_loop_test(tc, test_valid_name_gethostbyname3, 0, NELEMENTS(valid_names));
	tcase_add_loop_test(tc, test_valid_name_with_small_buffer, 1, 4);
	suite_add_tcase(s, tc);
	return s;
}

int main(void)
{
	int number_failed;
	Suite *s = libnss_juju_suite();
	SRunner *sr = srunner_create(s);
	srunner_set_tap(sr, "-");
	srunner_run_all(sr, CK_NORMAL);
	number_failed = srunner_ntests_failed(sr);
	srunner_free(sr);
	return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}


src/nss-juju.h
/* Copyright 2016 Canonical Ltd.
 *
 * Licensed under the AGPLv3, see LICENSE file for details.
 */

#ifndef _NSS_JUJU_H_
#define _NSS_JUJU_H_

#include <nss.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <errno.h>

extern enum nss_status _nss_juju_gethostbyname3_r
	(const char	 *name,
	 int		  af,
	 struct hostent	 *result,
	 char		 *buffer,
	 size_t		  buflen,
	 int		 *errnop,
	 int		 *herrnop,
	 int32_t	 *ttlp,
	 char		**canonp);

extern enum nss_status _nss_juju_gethostbyname2_r
	(const char	*name,
	 int		 af,
	 struct hostent *result,
	 char		*buffer,
	 size_t		 buflen,
	 int		*errnop,
	 int		*herrnop);

extern enum nss_status _nss_juju_gethostbyname_r
	(const char	*name,
	 struct hostent *result,
	 char		*buffer,
	 size_t		 buflen,
	 int		*errnop,
	 int		*herrnop);

#endif	/* _NSS_JUJU_H_ */


src/nss-juju.c
/* Copyright 2016 Canonical Ltd.
 *
 * Licensed under the AGPLv3, see LICENSE file for details.
 *
 * Name Service Switch module for resolving Juju host names.
 *
 * Author: Andrew McDermott <andrew.mcdermott@canonical.com>
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <unistd.h>
#include <netdb.h>
#include <regex.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <assert.h>

#include "nss-juju.h"

#define ALIGN(x, a)		__ALIGN_MASK(x, (typeof(x))(a) - 1)
#define __ALIGN_MASK(x, mask)	(((x) + (mask)) & ~(mask))
#define ALIGN_PTR(x)		ALIGN(x, sizeof(uint8_t *))

#define NELEMENTS(A) ((sizeof((A)) / sizeof((A))[0]))

#define FAMILY_LENGTH(X) ((X) == AF_INET6 ? return 64 : 16)

#define IPv4_OCTET "([[:digit:]]{1,3})"
#define IPv6_OCTET "([[:xdigit:]]{1,4})"

#define NAME_PREFIX "^juju-ip-"

#define IPv4_HOSTNAME_RE NAME_PREFIX					\
	IPv4_OCTET "-" IPv4_OCTET "-" IPv4_OCTET "-" IPv4_OCTET "$"

#define IPv6_HOSTNAME_RE NAME_PREFIX					\
	IPv6_OCTET "-" IPv6_OCTET "-" IPv6_OCTET "-" IPv6_OCTET "-"	\
	IPv6_OCTET "-" IPv6_OCTET "-" IPv6_OCTET "-" IPv6_OCTET "$"

struct _nss_juju_addrinfo {
	char ip_str[INET6_ADDRSTRLEN];
	int address_family;	/* addr discriminator */
	union {
		struct in_addr i4;
		struct in6_addr i6;
	} addr;
};

struct matcher {
	const char *pattern;
	const regex_t *pr;	/* non NULL if 're' is a valid RE. */
	regex_t re;
};

static struct matcher matchers[] = {
	{ IPv6_HOSTNAME_RE, NULL },
	{ IPv4_HOSTNAME_RE, NULL }
};

/*
 * shared library constructor
 */
static void __attribute__((constructor)) _init(void)
{
	size_t i;

	for (i = 0; i < NELEMENTS(matchers); i++) {
		if (regcomp(&matchers[i].re, matchers[i].pattern, REG_EXTENDED) == 0)
			matchers[i].pr = &matchers[i].re;
		assert(matchers[i].pr != NULL);
	}
}

/*
 * shared library destructor
 */
static void __attribute__((destructor)) _fini(void)
{
	size_t i;

	for (i = 0; i < NELEMENTS(matchers); i++) {
		if (matchers[i].pr != NULL) {
			regfree(&matchers[i].re);
			matchers[i].pr = NULL;
		}
	}
}

static int find_address(const regex_t *r,
			const char *hostname,
			struct _nss_juju_addrinfo *ai)
{
	size_t max_groups = 9;
	size_t n_groups = 0;
	regmatch_t groups[9];
	size_t i;
	int rc = regexec(r, hostname, max_groups, groups, 0);

	if (rc != 0)
		return -1;

	for (i = 0; i < max_groups; i++, n_groups++) {
		if (groups[i].rm_so == -1)
			break;
	}

	if (n_groups == 5) {
		sprintf(ai->ip_str, "%.*s.%.*s.%.*s.%.*s",
			groups[1].rm_eo - groups[1].rm_so, hostname + groups[1].rm_so,
			groups[2].rm_eo - groups[2].rm_so, hostname + groups[2].rm_so,
			groups[3].rm_eo - groups[3].rm_so, hostname + groups[3].rm_so,
			groups[4].rm_eo - groups[4].rm_so, hostname + groups[4].rm_so);

		if (inet_aton(ai->ip_str, &ai->addr.i4) != 0) {
			ai->address_family = AF_INET;
			return 0;
		}
	} else if (n_groups == 9) {
		sprintf(ai->ip_str, "%.*s:%.*s:%.*s:%.*s:%.*s:%.*s:%.*s:%.*s",
			groups[1].rm_eo - groups[1].rm_so, hostname + groups[1].rm_so,
			groups[2].rm_eo - groups[2].rm_so, hostname + groups[2].rm_so,
			groups[3].rm_eo - groups[3].rm_so, hostname + groups[3].rm_so,
			groups[4].rm_eo - groups[4].rm_so, hostname + groups[4].rm_so,
			groups[5].rm_eo - groups[5].rm_so, hostname + groups[5].rm_so,
			groups[6].rm_eo - groups[6].rm_so, hostname + groups[6].rm_so,
			groups[7].rm_eo - groups[7].rm_so, hostname + groups[7].rm_so,
			groups[8].rm_eo - groups[8].rm_so, hostname + groups[8].rm_so);

		if (inet_pton(AF_INET6, ai->ip_str, &ai->addr.i6) == 1) {
			ai->address_family = AF_INET6;
			return 0;
		}
	}

	return -1;
}

static int nss_juju_getaddrinfo(const char *name,
				struct _nss_juju_addrinfo *ai)
{
	size_t i;

	for (i = 0; i < NELEMENTS(matchers); i++) {
		if (matchers[i].pr == NULL)
			continue;
		if (find_address(matchers[i].pr, name, ai) == 0)
			return 0;
	}

	return -1;
}

enum nss_status _nss_juju_gethostbyname3_r(
	const char	 *name,
	int		  af,
	struct hostent	 *result,
	char		 *buffer,
	size_t		  buflen,
	int		 *errnop,
	int		 *herrnop,
	int32_t		 *ttlp,
	char		**canonp)
{
	size_t namelen = 0;
	struct _nss_juju_addrinfo ai;
	size_t bytes_required = 0;
	size_t buffer_offset = 0;
	char *aliases = NULL;
	char *addr_dst = NULL;
	char *addr_list = NULL;

	if (name == NULL) {
		*errnop = EINVAL;
		goto return_nss_status_unavail;
	}

	namelen = strlen(name);

	if (namelen == 0) {
		*errnop = ENOENT;
		goto return_nss_status_notfound;
	}

	if (buffer == NULL) {
		*errnop = EINVAL;
		goto return_nss_status_unavail;
	}

	ai.address_family = af;

	if (nss_juju_getaddrinfo(name, &ai) != 0) {
		*errnop = EADDRNOTAVAIL;
		goto return_nss_status_notfound;
	}

	/* name */
	bytes_required = ALIGN_PTR(namelen + 1);

	/* aliases */
	bytes_required += sizeof(char *);

	/* addresses */
	if (ai.address_family == AF_INET6) {
		bytes_required += ALIGN_PTR(sizeof ai.addr.i6);
	} else {
		bytes_required += ALIGN_PTR(sizeof ai.addr.i4);
	}

	/* null-terminated array of address pointers. */
	bytes_required += 2 * sizeof(char *);

	if (buflen < bytes_required) {
		*errnop = ERANGE;
		goto return_nss_status_tryagain;
	}

	/* Write name. */
	strcpy(buffer, name);
	result->h_name = buffer;
	buffer_offset = ALIGN_PTR(namelen + 1);

	/* Write aliases. */
	aliases = buffer + buffer_offset;
	*(char **)aliases = NULL;
	buffer_offset += sizeof(char *);
	result->h_aliases = (char **)aliases;

	/* Write address type and length. */
	result->h_addrtype = ai.address_family;
	result->h_length = ai.address_family == AF_INET6 ? sizeof ai.addr.i6 : sizeof ai.addr.i4;

	/* Write the single address. */
	addr_dst = buffer + buffer_offset;

	assert(ai.address_family == AF_INET6 || ai.address_family == AF_INET);

	if (ai.address_family == AF_INET6) {
		memcpy(addr_dst, &ai.addr.i6, result->h_length);
	} else {
		memcpy(addr_dst, &ai.addr.i4, result->h_length);
	}

	buffer_offset += ALIGN_PTR(result->h_length);

	/* Enumerate and terminate the addresses */
	addr_list = buffer + buffer_offset;
	((char **) addr_list)[0] = addr_dst;
	((char **) addr_list)[1] = NULL;
	buffer_offset += 2 * sizeof(char *);
	result->h_addr_list = (char **)addr_list;

	/* Did we write what we said we would? */
	assert(buffer_offset == bytes_required);

	if (ttlp != NULL)
		*ttlp = 0;

	if (canonp != NULL)
		*canonp = result->h_name;

	*errnop = 0;
	*herrnop = 0;
	return NSS_STATUS_SUCCESS;

return_nss_status_unavail:
	*herrnop = NO_DATA;
	return NSS_STATUS_UNAVAIL;

return_nss_status_tryagain:
	*herrnop = TRY_AGAIN;
	return NSS_STATUS_TRYAGAIN;

return_nss_status_notfound:
	*herrnop = HOST_NOT_FOUND;
	return NSS_STATUS_NOTFOUND;
}

enum nss_status _nss_juju_gethostbyname2_r(
	const char	*name,
	int		 af,
	struct hostent	*result,
	char		*buffer,
	size_t		 buflen,
	int		*errnop,
	int		*herrnop)
{
	return _nss_juju_gethostbyname3_r(name, af, result,
					  buffer, buflen,
					  errnop, herrnop,
					  NULL, NULL);
}

enum nss_status _nss_juju_gethostbyname_r(
	const char	*name,
	struct hostent	*result,
	char		*buffer,
	size_t		 buflen,
	int		*errnop,
	int		*herrnop)
{
	return _nss_juju_gethostbyname3_r(name, AF_UNSPEC, result,
					  buffer, buflen,
					  errnop, herrnop,
					  NULL, NULL);
}
